interpreter/cling/lib/Interpreter/DynamicLibraryManagerSymbol.cpp

```diff
@@ -555,355 +511,366 @@ void Dyld::BuildBloomFilter(LibraryPath* Lib,
-// can't use the "static" keyword because its address is used for
-// GetMainExecutable (since some platforms don't support taking the
-// address of main, and some platforms can't implement GetMainExecutable
-// without being given the address of a function in the main executable).
-std::string GetExecutablePath() {
-   // This just needs to be some symbol in the binary; C++ doesn't
-   // allow taking the address of ::main however.
-   return cling::DynamicLibraryManager::getSymbolLocation(&GetExecutablePath);
-}
+    if (DEBUG > 7)
+      cling::errs() << "Dyld::ResolveSymbol: Search found no match!\n";
+
+    /*
+      if (DEBUG > 7) {
+      cling::errs() << "Dyld::ResolveSymbol: Structs after ResolveSymbol:\n");
+
+      cling::errs() << "Dyld::ResolveSymbol - sPaths:\n");
+      size_t x = 0;
+      for (const auto &item : sPaths.GetPaths())
+      cling::errs() << "Dyld::ResolveSymbol << [" x++ << "]: " << item << "\n";
+
+      cling::errs() << "Dyld::ResolveSymbol - sLibs:\n");
+      x = 0;
+      for (const auto &item : sLibraries.GetLibraries())
+      cling::errs() << "Dyld::ResolveSymbol ["
+      << x++ << "]: " << item->Path << ", "
+      << item->LibName << "\n";
+
+      cling::errs() << "Dyld::ResolveSymbol - sSysLibs:");
+      x = 0;
+      for (const auto &item : sSysLibraries.GetLibraries())
+      cling::errs() << "Dyld::ResolveSymbol ["
+      << x++ << "]: " << item->Path << ", "
+      << item->LibName << "\n";
+
+      Info("Dyld::ResolveSymbol", "- sQueriedLibs:");
+      x = 0;
+      for (const auto &item : sQueriedLibraries)
+      cling::errs() << "Dyld::ResolveSymbol ["
+      << x++ << "]: " << item->Path << ", "
+      << item->LibName << "\n";
+      }
+    */
+
+    return ""; // Search found no match.
+  }
 
-namespace cling {
   void DynamicLibraryManager::initializeDyld(
-           std::function<bool(llvm::StringRef)> shouldPermanentlyIgnore) const {
-    assert(!s_IsDyldInitialized);
-    s_ShouldPermanentlyIgnoreCallback = shouldPermanentlyIgnore;
+                 std::function<bool(llvm::StringRef)> shouldPermanentlyIgnore) {
+    assert(!m_Dyld && "Already initialized!");
 
     std::string exeP = GetExecutablePath();
     auto ObjF =
       cantFail(llvm::object::ObjectFile::createObjectFile(exeP));
-       s_ExecutableFormat = ObjF.getBinary()->getFileFormatName();
 
-    s_IsDyldInitialized = true;
+    m_Dyld = new Dyld(*this, shouldPermanentlyIgnore,
+                      ObjF.getBinary()->getFileFormatName());
   }
 
   std::string
   DynamicLibraryManager::searchLibrariesForSymbol(const std::string& mangledName,
                                            bool searchSystem/* = true*/) const {
-    assert(s_IsDyldInitialized && "Must call initialize dyld before!");
-    static Dyld& dyld = Dyld::getInstance(*this);
-    return dyld.searchLibrariesForSymbol(mangledName, searchSystem);
+    assert(m_Dyld && "Must call initialize dyld before!");
```

Comment:
Can this not be guaranteed by the constructor of `DynamicLibraryManager`?

Reply:
No, we want this to be initialized on demand by users which actually need such a facility.

Reply:
Can one currently instantiate a DynamicLibraryManager without instantiating a Dyld? In which cases does that happen, even for cling?

Reply:
That's the point yes, cling standalone currently creates an object of type DynamicLibraryManager without any Dyld complications.